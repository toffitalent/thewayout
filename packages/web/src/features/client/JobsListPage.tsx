import { Box, Flex, Heading, Spinner } from '@disruptive-labs/ui';
import { useRouter } from 'next/router';
import { useCallback, useEffect, useRef } from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';
import { SEO } from '@app/components/SEO';
import { useAppDispatch, useAppSelector } from '@app/hooks';
import { selectAuthUser } from '../auth';
import { listAllJobs } from './actions';
import styles from './JobsListPage.module.scss';
import { ListItem } from './ListItem';
import { selectAllJobsList, selectTotalJobs } from './reducer';

export const JobsListPage = () => {
  const router = useRouter();
  const user = useAppSelector(selectAuthUser);
  const dispatch = useAppDispatch();
  const isFetching = useRef(false);
  const jobs = useAppSelector(selectAllJobsList);
  const total = useAppSelector(selectTotalJobs);

  useEffect(() => {
    if (user?.id) {
      isFetching.current = true;
      dispatch(listAllJobs({}))
        .unwrap()
        .then(() => {
          isFetching.current = false;
        });
    }
  }, [dispatch, user?.id]);

  const onClick = useCallback(
    (id: string) => router.push(`/client/jobs/${id}/`),
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [],
  );

  return (
    <>
      <SEO title="Job Postings" />
      <Flex>
        <Heading fontSize="3xl" mb={8}>
          Job Postings
        </Heading>

        <Flex className={styles.listContainer}>
          <InfiniteScroll
            dataLength={jobs.length}
            next={() => {
              if (!isFetching.current) {
                isFetching.current = true;
                if (user?.id) {
                  dispatch(listAllJobs({})).then(() => {
                    isFetching.current = false;
                  });
                }
              }
            }}
            hasMore={total > jobs.length}
            loader={
              <Box key={0} display="flex" justifyContent="center" mt={14}>
                <Spinner data-testid="loading-spinner" />
              </Box>
            }
          >
            <Flex width="full" container>
              {jobs.map((job) => (
                <Flex width="full" key={JSON.stringify(job)} mb={4}>
                  <ListItem {...job} onClick={onClick} />
                </Flex>
              ))}
            </Flex>
          </InfiniteScroll>
        </Flex>
      </Flex>
    </>
  );
};
