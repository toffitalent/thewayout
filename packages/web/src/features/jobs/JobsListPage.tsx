import { Box, Container, Flex, Heading, Spinner } from '@disruptive-labs/ui';
import { useRouter } from 'next/router';
import { useEffect, useRef } from 'react';
import InfiniteScroll from 'react-infinite-scroll-component';
import { selectAllJobsList, selectTotalJobs } from '@app/features/client';
import { listAllJobs } from '@app/features/client/actions';
import { useAppDispatch, useAppSelector } from '@app/hooks';
import styles from './JobsListPage.module.scss';
import { ListItem } from './ListItem';

export const JobsListPage = () => {
  const router = useRouter();
  const dispatch = useAppDispatch();
  const isFetching = useRef(false);

  const jobs = useAppSelector(selectAllJobsList);
  const total = useAppSelector(selectTotalJobs);

  useEffect(() => {
    isFetching.current = true;
    dispatch(listAllJobs({}))
      .unwrap()
      .then(() => {
        isFetching.current = false;
      });
  }, [dispatch]);

  const onClick = (id: string) => router.push(`/jobs/${id}/`);

  return (
    <Container>
      <Heading fontSize="2xl" mt={8} mb={8}>
        Current Opportunities
      </Heading>

      <Flex className={styles.listContainer}>
        <InfiniteScroll
          dataLength={jobs.length}
          next={() => {
            if (!isFetching.current) {
              isFetching.current = true;
              dispatch(listAllJobs({})).then(() => {
                isFetching.current = false;
              });
            }
          }}
          hasMore={total > jobs.length}
          loader={
            <Box key={0} display="flex" justifyContent="center" mt={14}>
              <Spinner data-testid="loading-spinner" />
            </Box>
          }
        >
          <Flex width="full" container>
            {jobs.map((job) => (
              <Flex width="full" key={JSON.stringify(job)} mb={4}>
                <ListItem {...job} onClick={onClick} />
              </Flex>
            ))}
          </Flex>
        </InfiniteScroll>
      </Flex>
    </Container>
  );
};
