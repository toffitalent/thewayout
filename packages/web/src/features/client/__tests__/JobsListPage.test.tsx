import { fixtures, mockState, render } from '@test';
import { JobsListPage } from '../JobsListPage';

const jobToSnap = {
  ...fixtures.job,
  createdAt: new Date(),
};

describe('JobsListPage', () => {
  test('renders correctly', async () => {
    const { container } = render(<JobsListPage />, {
      initialState: {
        clientJobs: {
          jobsList: {
            ids: [jobToSnap.id],
            entities: { [jobToSnap.id]: jobToSnap },
            total: 1,
            listUpdatedAt: '',
          },
        },
      },
    });
    expect(container).toMatchSnapshot();
  });
});
