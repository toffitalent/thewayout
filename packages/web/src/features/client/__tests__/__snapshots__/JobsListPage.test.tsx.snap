// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobsListPage renders correctly 1`] = `
<div>
  <div
    class=""
  >
    <h2
      class="text marginBottom-8 fontSize-3xl"
    >
      Job Postings
    </h2>
    <div
      class="listContainer"
    >
      <div
        class="infinite-scroll-component__outerdiv"
      >
        <div
          class="infinite-scroll-component "
          style="height: auto; overflow: auto;"
        >
          <div
            class="container width-full"
          >
            <div
              class="width-full marginBottom-4"
            >
              <div
                class="listItem borderBottom paddingTop-5"
              >
                <div
                  class="flexDirection-row marginBottom-5"
                >
                  <div
                    class="icon rounded bgcolor-orange-200 marginRight-5"
                  >
                    <svg
                      aria-hidden="true"
                      class="icon color-orange-800 width-6 height-6"
                      focusable="false"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M0 0h24v24H0z"
                        stroke="none"
                      />
                      <path
                        d="M3 9a2 2 0 012-2h14a2 2 0 012 2v9a2 2 0 01-2 2H5a2 2 0 01-2-2zM8 7V5a2 2 0 012-2h4a2 2 0 012 2v2M12 12v.01M3 13a20 20 0 0018 0"
                      />
                    </svg>
                  </div>
                  <div
                    class=""
                  >
                    <p
                      class="text fontSize-xl fontWeight-700"
                    >
                      Car Wash Attendant
                    </p>
                    <div
                      class="color-grey-600 flexWrap-wrap display-flex fontSize-sm"
                    >
                      <p
                        class="text"
                      >
                        Hybrid (Full-time)
                      </p>
                      <p
                        class="text"
                      >
                         • 
                        $15 - $30/hour
                      </p>
                      <p
                        class="text"
                      >
                         • 
                        0 Applicants
                      </p>
                    </div>
                    <p
                      class="text color-grey-600 paddingTop-4 paddingBottom-4"
                    >
                      Wash attendants are responsible for performing the daily operational practices of running the car wash facility.
                    </p>
                    <div
                      class="container spacing-2"
                    >
                      <div
                        class="item"
                      >
                        <div
                          class="badge borderColor-blue-200 bgcolor-blue-200 color-blue-800"
                        >
                          NO DRUGS OFFENSES
                        </div>
                      </div>
                      <div
                        class="item"
                      >
                        <div
                          class="badge borderColor-blue-200 bgcolor-blue-200 color-blue-800"
                        >
                          NO MOTOR VEHICLE OFFENSES
                        </div>
                      </div>
                      <div
                        class="item"
                      >
                        <div
                          class="badge borderColor-blue-200 bgcolor-blue-200 color-blue-800"
                        >
                          NO SEXUAL OFFENSES
                        </div>
                      </div>
                      <div
                        class="item"
                      >
                        <div
                          class="badge borderColor-blue-200 bgcolor-blue-200 color-blue-800"
                        >
                          NO VIOLENT OFFENSES
                        </div>
                      </div>
                      <div
                        class="item"
                      >
                        <div
                          class="badge borderColor-blue-200 bgcolor-blue-200 color-blue-800"
                        >
                          NO THEFT OFFENSES
                        </div>
                      </div>
                      <div
                        class="item"
                      >
                        <div
                          class="badge borderColor-blue-200 bgcolor-blue-200 color-blue-800"
                        >
                          NO WHITE - COLLAR OFFENSES
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
